File: 02_devApplication_IOS_avec_Objective_C/07_MKMapItem/07_MKMapItem/ViewController.m
//
//  ViewController.m
//  07_MKMapItem
//
//  Created by etudiant on 4/20/17.
//  Copyright © 2017 etudiant. All rights reserved.
//

#import "ViewController.h"
#import <CoreLocation/CoreLocation.h>
#import <MapKit/MapKit.h>
#import <Contacts/Contacts.h>
@interface ViewController ()
@property (weak, nonatomic) IBOutlet UITextField *txtAdresse;
@property (weak, nonatomic) IBOutlet UITextField *txtVille;
@property (weak, nonatomic) IBOutlet UITextField *txtCP;
@property (weak, nonatomic) IBOutlet UITextField *txtPays;

@property(nonatomic) CLLocationCoordinate2D coordonnesDestination;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
}

- (IBAction)btnAfficherTrajetTouched:(id)sender
{
    
    //recuper une instance du service de geolocalisation
    CLGeocoder * geocoder = [[CLGeocoder alloc]init];
    
    //crée une chaine contenant l'adresse de destination
    NSString * adresse = [NSString stringWithFormat:@"%@ %@ %@ %@",self.txtAdresse.text,self.txtVille.text,self.txtCP.text,self.txtPays.text];
    NSLog(@"%@",adresse);
    
    //Passer l'adresse obtenue à partir de la saisie au service de géolocalisation
    [geocoder geocodeAddressString:adresse
                completionHandler:^(NSArray<CLPlacemark *> * _Nullable placemarks, NSError * _Nullable error)
                {
                    //analyser l'argument arror pour savoir si l'adresse saisie a été trouvée
                    if(error)
                    {
                        NSLog(@"Echec de service geocode avec erreur %@",error);
                        return;
                    }
                    //si geocoder a trouver une adresse au plusieurs
                    //qui corresponde à la saisie données,il les met dans un tableau qui nous est forni dans l'argument placemarks
                    //si aucune adresse n'a été trouvé placemarks contient 'nil'
                    if(placemarks || placemarks.count > 0)
                    {
                        //choisir la prémier adresse trouvée
                       CLPlacemark * placemark = placemarks[0];
                        //récupere les coordonnée géographique de cette adresse
                       float latitude = placemark.location.coordinate.latitude;
                        float longitude = placemark.location.coordinate.longitude;
                        NSLog(@"latitude: %f, longutude: %f",latitude,longitude);
                        
                        self.coordonnesDestination = placemark.location.coordinate;
                        
                        [self afficherMap];
                        
                    }
                }];
}
-(void)afficherMap
{
    //créer un dictionnaire qui doit contenir les informations saisie c'est à dire les information du point d'arrivée de notre trajet
    //Ce dictionnaire possede des cles specifique
    NSDictionary * dicoDestination = @{
                        CNPostalAddressCityKey:self.txtVille.text,
                        CNPostalAddressStreetKey:self.txtAdresse.text,
                        CNPostalAddressPostalCodeKey:self.txtCP.text,
                        CNPostalAddressCountryKey:self.txtPays.text
                                       };
    MKPlacemark * placemarkDestination = [[MKPlacemark alloc]initWithCoordinate:self.coordonnesDestination addressDictionary:dicoDestination];
    MKMapItem * mkMapItem = [[MKMapItem alloc]initWithPlacemark:placemarkDestination];
    
    //preciser des eventuelles options de parcours ces options doivent etre données aussi dans un dictionnaire avec des cles specifique
    NSDictionary * dicoOptions = @{
                                MKLaunchOptionsDirectionsModeKey :
                                MKLaunchOptionsDirectionsModeDriving
                                   };
    [mkMapItem openInMapsWithLaunchOptions:dicoOptions];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: 04_devApplication_IOS_avec_Swift/08_/08_/ViewController.swift
//
//  ViewController.swift
//  08_
//
//  Created by etudiant on 5/12/17.
//  Copyright © 2017 etudiant. All rights reserved.
//

import UIKit
import MapKit
import Contacts
import CoreLocation

class ViewController: UIViewController {

    @IBOutlet weak var txtAdress: UITextField!
    @IBOutlet weak var txtVille: UITextField!
    @IBOutlet weak var txtCp: UITextField!
    @IBOutlet weak var txtPays: UITextField!
    
    var coordonneDestination:CLLocationCoordinate2D?
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
    }
    
    func afficherMap() {
        //la carte avec la localisation courante, le point d'arrive et le trajet sera afficher par l'api Maps
        //cette api a besoin de lui passer les information en utilisant un dico contenant des clef qu'elle connait
        //Créer dico avec ces clef specifique et avec les valeur saisies
        
        let dicoAdresseDestination = [CNPostalAddressStreetKey: txtAdress.text!,
                                      CNPostalAddressCityKey: txtVille.text!,
                                      CNPostalAddressPostalCodeKey: txtCp.text!,
                                      CNPostalAddressCountryKey: txtPays.text!]
        
        //reunir les coordoner trouver par le geocoder et les saisies dans un seul objet
        let placemarkDestination = MKPlacemark(coordinate: coordonneDestination!,
                                               addressDictionary: dicoAdresseDestination)
        //creer un objet de type MKMapItem
        let mkMapItem = MKMapItem(placemark: placemarkDestination)
        
        //preciser des eventuelles option de parcours
        //ces options doivent etre donner aussi dans un dico avec des clef sprecifique
        let options = [MKLaunchOptionsDirectionsModeKey:MKLaunchOptionsDirectionsModeDriving]
        
        //afficher la map avec le parcour calculer en fct des coordonner du point final et des options choisies
        //ce parcours aura comme point de depart la position courant de l'utilisateur
        mkMapItem.openInMaps(launchOptions: options)
        
    }

    @IBAction func afficherTrajetTouch(_ sender: Any) {
        //creer a partir des donner saisie, une chaine de caractere creant l'adress ou on veut arriver
        let adress = "\(txtAdress.text!), \(txtCp.text!), \(txtVille.text!), \(txtPays.text!)"
        
        //pour interroger le service de geolocalisation distant, je dois utiliser un objet CLGeocoder
        let geocoder = CLGeocoder()
        geocoder.geocodeAddressString(adress){ (tbPlacemarks, erreur) in
            if erreur != nil{
                print("Code erreur geocoder: \(erreur!.localizedDescription)")
            } else if tbPlacemarks != nil && tbPlacemarks!.count > 0 {
                //afficher le premier item du tableau placemarks
                let placemark = tbPlacemarks![0]
                let location = placemark.location
                self.coordonneDestination = location!.coordinate
                
                self.afficherMap()
                print("Les coordonées du premier point trouvé")
                print("Latitude: \(self.coordonneDestination!.latitude), longitude: \(self.coordonneDestination!.longitude)")
            }
        }
    }

}

